pkg/cgroups/manager/manager.go

```diff
@@ -0,0 +1,405 @@
+		}
+		// The container ID field is of the form: <container-runtime>://<containerID>
+		// Example:containerd://e275d1a37782ab30008aa3ae6666cccefe53b3a14a2ab5a8dc459939107c8c0e
+		_, after, found := strings.Cut(cId, "//")
+		if !found || after == "" {
+			log.WithFields(logrus.Fields{
+				logfields.K8sPodName:   pod.Name,
+				logfields.K8sNamespace: pod.Namespace,
+				"container-id":         cId,
+			}).Error("unexpected container ID")
+			continue
+		}
+		cId = after
+		if _, ok := pm.containers[cId]; ok {
+			currContainers[cId] = struct{}{}
+			// Container cgroup path doesn't change as long as the container id
+			// is the same.
+			continue
+		}
+		pm.containers[cId] = struct{}{}
+		currContainers[cId] = struct{}{}
+
+		// Container could've been gone, so don't log any errors.
+		cgrpPath, err := m.getContainerCgroupPath(id, cId, pod.Status.QOSClass)
+		if err != nil {
+			log.WithFields(logrus.Fields{
+				logfields.K8sPodName:   pod.Name,
+				logfields.K8sNamespace: pod.Namespace,
+				"container-id":         cId,
+			}).WithError(err).Debugf("failed to get container metadata")
+			continue
+		}
+		cgrpId, err := m.cgroupsChecker.GetCgroupID(cgrpPath)
+		if err != nil {
+			log.WithFields(logrus.Fields{
+				logfields.K8sPodName:   pod.Name,
+				logfields.K8sNamespace: pod.Namespace,
+				"cgroup-path":          cgrpPath,
+			}).WithError(err).Debugf("failed to get cgroup id")
+			continue
+		}
+		m.containerMetadataByCgrpId[cgrpId] = &containerMetadata{
+			cgroupId:   cgrpId,
+			cgroupPath: cgrpPath,
+			podId:      id,
+		}
+	}
+	// Clean up any pod's old containers.
+	if oldPod != nil {
+		for _, c := range oldPod.Status.ContainerStatuses {
+			// Pod status fields other than containers can be updated so check for
+			// containers that were deleted.
+			if _, ok := currContainers[c.ContainerID]; !ok {
+				delete(pm.containers, c.ContainerID)
+			}
+		}
+	}
+}
+
+func (m *CgroupManager) deletePodMetadata(pod *v1.Pod) {
+	podId := string(pod.ObjectMeta.UID)
+
+	if _, ok := m.podMetadataById[podId]; !ok {
+		return
+	}
+	for k, cm := range m.containerMetadataByCgrpId {
+		if cm.podId == podId {
+			delete(m.containerMetadataByCgrpId, k)
+		}
+	}
+	delete(m.podMetadataById, podId)
+}
+
+func (m *CgroupManager) getPodMetadata(cgroupId uint64, podMetadataOut chan *PodMetadata) {
+	cm, ok := m.containerMetadataByCgrpId[cgroupId]
+	if !ok {
+		log.WithFields(logrus.Fields{
+			"container-cgroup-id": cgroupId,
+		}).Debugf("Metadata not found")
+		close(podMetadataOut)
+		return
+	}
+
+	pm, ok := m.podMetadataById[cm.podId]
+	if !ok {
+		log.WithFields(logrus.Fields{
+			"container-cgroup-id": cgroupId,
+		}).Debugf("Pod metadata not found")
+		close(podMetadataOut)
+		return
+	}
+	podMetadata := PodMetadata{
+		Name:      pm.name,
+		Namespace: pm.namespace,
+	}
+	podMetadata.IPs = append(podMetadata.IPs, pm.ips...)
+	log.WithFields(logrus.Fields{
+		"container-cgroup-id": cgroupId,
+	}).Debugf("Pod metadata: %+v", podMetadata)
```

Comment:
Hey @aditighag how frequently would you expect this log to be triggered? I'm noticing it 1000+ times in the first 5 minutes of Cilium running, which seems a bit high. Given the channel usage and so on this might be introducing more memory churn / pressure than necessary. Is it expected? Many of the logs seem to repeat for the exact same Pods over and again.

Reply:
This log was added before we had visibility via CLI (https://github.com/cilium/cilium/pull/23641). We can remove the log now.

Reply:
The function is called by Hubble every time a `trace-sock` is handled, that might explain the high volume. I did not realize that `GetPodMetadataForContainer` did a channel allocation and communication for each call, that might actually be a performance problem :confused: 

Hubble itself will not create any Hubble flow if the socket is not associated with a pod (i.e. if this function here returns nil), but maybe we should think about also reducing the amount of events emitted by the datapath if that's somehow possible.

Reply:
I vaguely remember having this discussion with Sebastian on whether Hubble cached metadata results. The answer was no as Hubble also retrieves metadata from other subsystems such as ipcache, k8s watcher, etc. Please correct me if that's not the case.

I don't have cycles at the moment, but at some point I would like to revisit this: some of the options being aggregating events at the datapath, or having a cache in the cgroup manager. FWIW, events are being generated at the socket and not packet level, but I agree there is a scope for optimization. Btw the aggregation task is part of the follow-ups meta issue - https://github.com/cilium/cilium/issues/21510.

Reply:
Here is a PR that adds caching - https://github.com/cilium/cilium/pull/32615.